### ISSUE TITLE
heap-buffer-overflow in _WM_do_note_off

### ISSUE THREAD

**Issue by gy741** [Mindwerks/wildmidi#180]
Hi.

I found a heap-buffer-overflow in wildmidi.

Please confirm.

Thanks.

Summary: heap-buffer-overflow
OS: Ubuntu 16.04 64bit
wildmidi version: WildMidi 0.4.2(ec0e2880b05c62562db16fbe448bafd19adb4593)
Steps to reproduce:
1.Download the .POC files.
2.Compile the source code with ASan or Run wildmidi as valgrind.
3.Execute the following command
: ./wildmidi $PoC -o /dev/null
PoC download : [PoC](https://github.com/gy741/PoC/raw/master/000005_PoC_heap_wildmidi)

```
valgrind log
-------------
==20068== Conditional jump or move depends on uninitialised value(s) |
==20068==    at 0x4E529A2: _WM_do_note_off (internal_midi.c:627)
==20068==    by 0x4E4089E: WM_GetOutput_Linear (wildmidi_lib.c:858)
==20068==    by 0x4E4089E: WildMidi_GetOutput (wildmidi_lib.c:1947)
==20068==    by 0x403DC1: main (wildmidi.c:2017)
-------------
ASan log
-------------
==16176==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7fc9f61f118e at pc 0x7fc9f5eecdc4 bp 0x7ffce4113d10 sp 0x7ffce4113d08
READ of size 1 at 0x7fc9f61f118e thread T0
#0 0x7fc9f5eecdc3 in _WM_do_note_off /root/karas/wildmidi/src/internal_midi.c:627:19
#1 0x7fc9f5ecbc00 in WM_GetOutput_Linear /root/karas/wildmidi/src/wildmidi_lib.c:856:17
#2 0x7fc9f5ecbc00 in WildMidi_GetOutput /root/karas/wildmidi/src/wildmidi_lib.c:1953
#3 0x50ec69 in main /root/karas/wildmidi/src/wildmidi.c:2017:19
#4 0x7fc9f4fb382f in __libc_start_main /build/glibc-bfm8X4/glibc-2.23/csu/../csu/libc-start.c:291
#5 0x41a698 in _start (/root/karas/wildmidi/wildmidi+0x41a698)
```

**Comment by sezero** [comment]
Confirmed.
```
                              [    ] [100] [ 0m 41s Processed] [69%] |
==9424== Invalid read of size 1
==9424==    at 0x805A909: _WM_do_note_off (internal_midi.c:627)
==9424==    by 0x804E0DD: WM_GetOutput_Linear (wildmidi_lib.c:856)
==9424==    by 0x80502E6: WildMidi_GetOutput (wildmidi_lib.c:1953)
==9424==    by 0x804B7E5: main (wildmidi.c:2017)
==9424==  Address 0x4073162 is 1,002 bytes inside a block of size 898,642 free'd
==9424==    at 0x400590A: free (vg_replace_malloc.c:323)
==9424==    by 0x8062A20: _WM_load_gus_pat (gus_pat.c:908)
==9424==    by 0x805F737: _WM_load_sample (sample.c:126)
==9424==    by 0x805F22D: _WM_load_patch (patches.c:84)
==9424==    by 0x805DA73: _WM_initMDI (internal_midi.c:1887)
==9424==    by 0x8054C3C: _WM_ParseNewMidi (f_midi.c:146)
==9424==    by 0x804FA01: WildMidi_Open (wildmidi_lib.c:1669)
==9424==    by 0x804AEAC: main (wildmidi.c:1804)
==9424== 
==9424== Invalid read of size 1
==9424==    at 0x805A93E: _WM_do_note_off (internal_midi.c:636)
==9424==    by 0x804E0DD: WM_GetOutput_Linear (wildmidi_lib.c:856)
==9424==    by 0x80502E6: WildMidi_GetOutput (wildmidi_lib.c:1953)
==9424==    by 0x804B7E5: main (wildmidi.c:2017)
==9424==  Address 0x4073160 is 1,000 bytes inside a block of size 898,642 free'd
==9424==    at 0x400590A: free (vg_replace_malloc.c:323)
==9424==    by 0x8062A20: _WM_load_gus_pat (gus_pat.c:908)
==9424==    by 0x805F737: _WM_load_sample (sample.c:126)
==9424==    by 0x805F22D: _WM_load_patch (patches.c:84)
==9424==    by 0x805DA73: _WM_initMDI (internal_midi.c:1887)
==9424==    by 0x8054C3C: _WM_ParseNewMidi (f_midi.c:146)
==9424==    by 0x804FA01: WildMidi_Open (wildmidi_lib.c:1669)
==9424==    by 0x804AEAC: main (wildmidi.c:1804)
==9424== 
==9424== Invalid write of size 1
==9424==    at 0x805A700: _WM_do_note_off_extra (internal_midi.c:568)
==9424==    by 0x805A967: _WM_do_note_off (internal_midi.c:642)
==9424==    by 0x804E0DD: WM_GetOutput_Linear (wildmidi_lib.c:856)
==9424==    by 0x80502E6: WildMidi_GetOutput (wildmidi_lib.c:1953)
==9424==    by 0x804B7E5: main (wildmidi.c:2017)
==9424==  Address 0x4073174 is 1,020 bytes inside a block of size 898,642 free'd
==9424==    at 0x400590A: free (vg_replace_malloc.c:323)
==9424==    by 0x8062A20: _WM_load_gus_pat (gus_pat.c:908)
==9424==    by 0x805F737: _WM_load_sample (sample.c:126)
==9424==    by 0x805F22D: _WM_load_patch (patches.c:84)
==9424==    by 0x805DA73: _WM_initMDI (internal_midi.c:1887)
==9424==    by 0x8054C3C: _WM_ParseNewMidi (f_midi.c:146)
==9424==    by 0x804FA01: WildMidi_Open (wildmidi_lib.c:1669)
==9424==    by 0x804AEAC: main (wildmidi.c:1804)
==9424== 
==9424== Invalid read of size 1
==9424==    at 0x805A707: _WM_do_note_off_extra (internal_midi.c:570)
==9424==    by 0x805A967: _WM_do_note_off (internal_midi.c:642)
==9424==    by 0x804E0DD: WM_GetOutput_Linear (wildmidi_lib.c:856)
==9424==    by 0x80502E6: WildMidi_GetOutput (wildmidi_lib.c:1953)
==9424==    by 0x804B7E5: main (wildmidi.c:2017)
==9424==  Address 0x4073160 is 1,000 bytes inside a block of size 898,642 free'd
==9424==    at 0x400590A: free (vg_replace_malloc.c:323)
==9424==    by 0x8062A20: _WM_load_gus_pat (gus_pat.c:908)
==9424==    by 0x805F737: _WM_load_sample (sample.c:126)
==9424==    by 0x805F22D: _WM_load_patch (patches.c:84)
==9424==    by 0x805DA73: _WM_initMDI (internal_midi.c:1887)
==9424==    by 0x8054C3C: _WM_ParseNewMidi (f_midi.c:146)
==9424==    by 0x804FA01: WildMidi_Open (wildmidi_lib.c:1669)
==9424==    by 0x804AEAC: main (wildmidi.c:1804)
==9424== 
==9424== Invalid read of size 1
==9424==    at 0x805A719: _WM_do_note_off_extra (internal_midi.c:571)
==9424==    by 0x805A967: _WM_do_note_off (internal_midi.c:642)
==9424==    by 0x804E0DD: WM_GetOutput_Linear (wildmidi_lib.c:856)
==9424==    by 0x80502E6: WildMidi_GetOutput (wildmidi_lib.c:1953)
==9424==    by 0x804B7E5: main (wildmidi.c:2017)
==9424==  Address 0x4073160 is 1,000 bytes inside a block of size 898,642 free'd
==9424==    at 0x400590A: free (vg_replace_malloc.c:323)
==9424==    by 0x8062A20: _WM_load_gus_pat (gus_pat.c:908)
==9424==    by 0x805F737: _WM_load_sample (sample.c:126)
==9424==    by 0x805F22D: _WM_load_patch (patches.c:84)
==9424==    by 0x805DA73: _WM_initMDI (internal_midi.c:1887)
==9424==    by 0x8054C3C: _WM_ParseNewMidi (f_midi.c:146)
==9424==    by 0x804FA01: WildMidi_Open (wildmidi_lib.c:1669)
==9424==    by 0x804AEAC: main (wildmidi.c:1804)
==9424== 
==9424== Invalid write of size 4
==9424==    at 0x805A73C: _WM_do_note_off_extra (internal_midi.c:574)
==9424==    by 0x805A967: _WM_do_note_off (internal_midi.c:642)
==9424==    by 0x804E0DD: WM_GetOutput_Linear (wildmidi_lib.c:856)
==9424==    by 0x80502E6: WildMidi_GetOutput (wildmidi_lib.c:1953)
==9424==    by 0x804B7E5: main (wildmidi.c:2017)
==9424==  Address 0x4073154 is 988 bytes inside a block of size 898,642 free'd
==9424==    at 0x400590A: free (vg_replace_malloc.c:323)
==9424==    by 0x8062A20: _WM_load_gus_pat (gus_pat.c:908)
==9424==    by 0x805F737: _WM_load_sample (sample.c:126)
==9424==    by 0x805F22D: _WM_load_patch (patches.c:84)
==9424==    by 0x805DA73: _WM_initMDI (internal_midi.c:1887)
==9424==    by 0x8054C3C: _WM_ParseNewMidi (f_midi.c:146)
==9424==    by 0x804FA01: WildMidi_Open (wildmidi_lib.c:1669)
==9424==    by 0x804AEAC: main (wildmidi.c:1804)
[...]
==9424== 
==9424== ERROR SUMMARY: 6 errors from 6 contexts (suppressed: 13 from 1)
==9424== malloc/free: in use at exit: 20 bytes in 1 blocks.
```

When the issue is hit, data->data.value>>8 is 149

**Comment by sezero** [comment]
When the issue is hit, data->data.value>>8 is 149
mdi->note_table is struct _note note_table[2][16][128] so the value > 128 is the problem, of course..

**Comment by sezero** [comment]
I can do the following to silently bound the note to 0..12.
However, someone with better midi knowledge than me must review this.
@chrisisonwildcode? @psi29a?

```
diff --git a/src/internal_midi.c b/src/internal_midi.c
index af99c18..7b57fe7 100644
--- a/src/internal_midi.c
+++ b/src/internal_midi.c
@@ -1474,6 +1474,7 @@ int _WM_midi_setup_noteoff(struct _mdi *mdi, uint8_t channel,
                            uint8_t note, uint8_t velocity) {
     MIDI_EVENT_DEBUG(__FUNCTION__,channel, note);
     _WM_CheckEventMemoryPool(mdi);
+    note &= 0x7f; /* silently bound note to 0..127 */
     mdi->events[mdi->event_count].do_event = *_WM_do_note_off;
     mdi->events[mdi->event_count].event_data.channel = channel;
     mdi->events[mdi->event_count].event_data.data.value = (note << 8) | velocity;
@@ -1486,6 +1487,7 @@ static int midi_setup_noteon(struct _mdi *mdi, uint8_t channel,
                              uint8_t note, uint8_t velocity) {
     MIDI_EVENT_DEBUG(__FUNCTION__,channel, note);
     _WM_CheckEventMemoryPool(mdi);
+    note &= 0x7f; /* silently bound note to 0..127 */
     mdi->events[mdi->event_count].do_event = *_WM_do_note_on;
     mdi->events[mdi->event_count].event_data.channel = channel;
     mdi->events[mdi->event_count].event_data.data.value = (note << 8) | velocity;
@@ -1501,6 +1503,7 @@ static int midi_setup_aftertouch(struct _mdi *mdi, uint8_t channel,
                                  uint8_t note, uint8_t pressure) {
     MIDI_EVENT_DEBUG(__FUNCTION__,channel, note);
     _WM_CheckEventMemoryPool(mdi);
+    note &= 0x7f; /* silently bound note to 0..127 */
     mdi->events[mdi->event_count].do_event = *_WM_do_aftertouch;
     mdi->events[mdi->event_count].event_data.channel = channel;
     mdi->events[mdi->event_count].event_data.data.value = (note << 8) | pressure;
```

**Comment by sezero** [comment]
I can do the following to silently bound the note to 0..127.
However, someone with better midi knowledge than me must review this.
@chrisisonwildcode? @psi29a?
```
diff --git a/src/internal_midi.c b/src/internal_midi.c
index af99c18..cca1480 100644
--- a/src/internal_midi.c
+++ b/src/internal_midi.c
@@ -1473,8 +1473,9 @@
 int _WM_midi_setup_noteoff(struct _mdi *mdi, uint8_t channel,
                            uint8_t note, uint8_t velocity) {
     MIDI_EVENT_DEBUG(__FUNCTION__,channel, note);
     _WM_CheckEventMemoryPool(mdi);
+    note &= 0x7f; /* silently bound note to 0..127 (github bug #180) */
     mdi->events[mdi->event_count].do_event = *_WM_do_note_off;
     mdi->events[mdi->event_count].event_data.channel = channel;
     mdi->events[mdi->event_count].event_data.data.value = (note << 8) | velocity;
     mdi->events[mdi->event_count].samples_to_next = 0;
@@ -1485,8 +1486,9 @@
 static int midi_setup_noteon(struct _mdi *mdi, uint8_t channel,
                              uint8_t note, uint8_t velocity) {
     MIDI_EVENT_DEBUG(__FUNCTION__,channel, note);
     _WM_CheckEventMemoryPool(mdi);
+    note &= 0x7f; /* silently bound note to 0..127 (github bug #180) */
     mdi->events[mdi->event_count].do_event = *_WM_do_note_on;
     mdi->events[mdi->event_count].event_data.channel = channel;
     mdi->events[mdi->event_count].event_data.data.value = (note << 8) | velocity;
     mdi->events[mdi->event_count].samples_to_next = 0;
@@ -1500,8 +1502,9 @@
 static int midi_setup_aftertouch(struct _mdi *mdi, uint8_t channel,
                                  uint8_t note, uint8_t pressure) {
     MIDI_EVENT_DEBUG(__FUNCTION__,channel, note);
     _WM_CheckEventMemoryPool(mdi);
+    note &= 0x7f; /* silently bound note to 0..127 (github bug #180) */
     mdi->events[mdi->event_count].do_event = *_WM_do_aftertouch;
     mdi->events[mdi->event_count].event_data.channel = channel;
     mdi->events[mdi->event_count].event_data.data.value = (note << 8) | pressure;
     mdi->events[mdi->event_count].samples_to_next = 0;
```

**Comment by psi29a** [comment]
I assume the input data is crafted and not apart of the MIDI spec... thus bounded it to 0..127 is a good thing to do.

@chrisisonwildcode: you agree?

**Comment by sezero** [comment]
PING?

**Comment by psi29a** [comment]
Bueller... Bueller...

**Comment by sezero** [comment]
Fixed using the patch above, for now. Closing.

If any corrections are to be made or anyone has a better solution, please reopen this.

**Comment by psi29a** [comment]
Moving on... thanks @sezero .

**Comment by sezero** [comment]
Backported the fix to wildmidi-0.3: commit 1878e981d
